src/ripple/app/tx/impl/PayChan.cpp

```diff
@@ -271,6 +283,13 @@ PayChanCreate::doApply()
     (*slep)[~sfSourceTag] = ctx_.tx[~sfSourceTag];
     (*slep)[~sfDestinationTag] = ctx_.tx[~sfDestinationTag];
 
+    if (ctx_.view().rules().enabled(fixPayChanV1))
+    {
+        auto const closeTime = ctx_.view().info().parentCloseTime;
+        if (ctx_.tx[~sfCancelAfter] && after(closeTime, ctx_.tx[sfCancelAfter]))
+            return temBAD_EXPIRATION;
```

Comment:
Shouldn't `tem` codes should only be used in `preflight`?

Reply:
You're probably right. There are instances of tem codes appearing in `preclaim` and `doApply` but if thats the general standard than I can make the change.

Reply:
I don't believe we have good documentation for when to return different ter codes, and we should write some. In the meantime, here are my thoughts:

A transactions has three major phases: 

1) Preflight. This is cheap. It happens before signature checks (which are expensive) and cannot read the ledger. In order to prevent broadcasting transactions to the network that fail signature checks, and therefore can't claim a fee, preflight is not allowed to return `tec` codes. (Note: there are several places where we check signatures: preflight2, invoke_preclaim, and in network ops; But not allowing `preflight` to return `tec` codes was modivated by the signature checks happening after preflight).

2) Preclaim. This is more expensive. This happens after signature checks (and after preflight) and has read access to the ledger.

3) Apply. This happens after preclaim and has read/write access to the ledger.

Two important points are:

1) Preflight is cheap and happens before signature checking.

2) If something returns a tem code it is not broadcast to the network. If something returns a tec code it is broadcast.

There are places in the code that return `tem` codes after `preflight`. The bad thing about this is someone can try to overload an individual server by making it do relatively expensive operations (like signature checks) and never pay fees for the work. The good thing is it saves the other servers on the network from doing the work because it is never broadcast.

However, I don't see a way to prevent someone from submitting bad signatures to an individual server and making that server do the same work that a transaction with a valid signature that returns a `tem` code would do. 

It's important to note that `tem` codes should only be returned for transactions that can't succeed in any ledger. For example, if an account has an insufficient balance in the current open ledger, that transaction should return a `tec` code and should still be broadcast. In some cases we may want to return `tec` codes even when a `tem` would to so the transaction is recorded in a ledger and can easily be retrieved and replayed to diagnose errors.

Given that returning a `tec` code would not prevent any attacks that I can see, and there are benefits to not broadcasting transactions, I don't think we want to blanket disallow returning tem codes from `preclaim` or `apply` or change the ones that are already there.

However, in most cases `preclaim` and `apply` should return `tec` codes.

Reply:
Just to add on to this, I am supportive of creating this fix amendment, but I believe the appropriate error code for this situation is probably `tecEXPIRED`:

> The transaction tried to create an object (such as an Offer or a Check) whose provided Expiration time has already passed.

It would not surprise me if EscrowCreate is inconsistent with this, too.

But here's why I think a `tec`-class code is appropriate:

1. The transaction isn't internally inconsistent or self-contradictory; it's only no good because there exists a previous ledger whose close time is higher than the specified expiration.
2. Since transactions can be delayed an indefinite amount of time between creation and confirmation¹, this error can occur for transactions that _were_ valid initially, got relayed to a bunch of servers (maybe queued), and then happened to have their expiration pass while they were awaiting confirmation. As a general rule, we want transactions to claim a fee (`tec` result code) if they've been relayed through the network, so transactions whose only fault is something is past expiration should use a `tec` code not a `tem` code.

¹ Some example situations that could cause a delay:
- Transaction was created and signed, saved, and then machine lost network connectivity or power. When it came back up an unspecified amount of time later, it submitted the transaction.
- Transaction was relayed to some servers during a period of high load, then fees went up and it didn't get relayed to enough servers to achieve consensus. Later on when load was lower some system retried the same transaction..
- The transaction was relayed through the network and on track to be validated, but then XRPL network had a temporary outage preventing new ledgers from being confirmed for a while (for example, half of validators became unable to talk to the other half due to a massive netsplit).

Reply:
Thank you both for the very very detailed explanation. This will help me a lot. I'm going to change the code to `tecEXPIRED`.